### ISSUE TITLE
Feature request: make normal mode agnostic of keyboard layout

### ISSUE THREAD

**Issue by max-baz** [mawww/kakoune#2151]
I have English and Russian keyboard layouts (with Latin and Cyrillic letters respectively), and kakoune is unusable in normal layout when Russian layout is active. To start typing text on a new line, I have to first switch to English layout, press "o", then switch to Russian layout and only then start typing the text. As you can imagine, this is very annoying.

Vim solves this using `langremap` configuration, where I can configure that in normal mode Cyrillic "щ" should correspond to Latin "o" (these keys are located in the same place on the physical keyboard), so I can press the "o" key in both English and Russian layouts and vim will do the same action.

This should also respect my own mappings, so if I configure a new mapping `map global user o :echo 'foo<ret>'`, this mapping should also be triggered when I press the "o" key in Russian layout (so the Cyrillic "щ") — I don't want to configure all of my mappings twice.

I'm not married to `langremap` approach, if you have better ideas how to support my scenario, I am happy to hear them.

**Comment by Screwtapello** [comment]
Because Kakoune is a terminal application, it only knows about Unicode characters, not about keys: it doesn't know what keyboard layout you're using, what physical key "щ" corresponds to, or what character that key would produce in English layout, so it's not really possible to solve this in a way that Just Works (although perhaps a GUI-based frontend could do this, someday).

As you've probably discovered, it's not possible to use `map` to do this, since `map` always maps directly to underlying commands, not to other mappings.

It might be useful for Kakoune to have an extra map "mode" that is always applied before the actual current mode, to cover this use-case:

    map global ovelay щ o
    map global user o ":echo foo<ret>"

Alternatively, a special map option whose keys are locale codes like `en_US` or `ru_RU`, and whose values are strings where each odd character gets automatically mapped to the following even character outside of insert mode. That way, Kakoune could ship with default config like:

    set-option -add global langmap "ru_RU=щo"

...and if Kakoune was started in a known locale, it would Just Work... assuming each locale has exactly one keyboard layout, or at least one popular keyboard layout, and nobody ever mapped anything to non-Latin keys.

**Comment by maximbaz** [comment]
I appreciate you desire to make things Just Work, but the assumptions you mentioned... I don't think we can expect them to hold 🙂 Maybe just follow the vim's approach then and make users define their langmaps without binding to a locale? We can have a Wiki page where people share their langmaps, for simple copy-paste.

Vim has two ways to configure `langmap`, either you define a list of "from-to" character pairs separated by comma, or you can group the characters in longer strings. This is my full `langmap` definition in vim:

```
set langmap+=ФИСВУАПРШОЛДЬТЩЗЙКЫЕГМЦЧНЯ;ABCDEFGHIJKLMNOPQRSTUVWXYZ " Cyrillic layout in normal mode
set langmap+=фисвуапршолдьтщзйкыегмцчня;abcdefghijklmnopqrstuvwxyz " Cyrillic layout in normal mode
set langmap+=ЖжЭэХхЪъ;\:\;\"\'{[}]                                 " Cyrillic layout in normal mode
```

**Comment by andreyorst** [comment]
Vim also supports this:
```vim
set keymap=russian-jcukenwin
set iminsert=0
set imsearch=0
```
Which means that you can switch layout within Vim only. And when you're in insert mode, you use desired layout, but if you're in normal mode, EN keymap is used. I was struggling with Russian layout too, and this fix works very well

**Comment by FlyingWombat** [comment]
I agree with the suggestion by @Screwtapello to have a special 'langmap' mode, that when activated is always applied over the current actual mode (except insert mode). This would be useful not only for language maps, but also for non-standard keyboard layouts, such as Dvorak.  

I'd also like to mention, such 'langmap' feature should keep recorded macros in mind. That way, the macros are still recorded in the editing language (instead of Cyrillic), and stay consistent whether 'langmap' is on or off.  

Would [Kakoune's design philosophy](https://github.com/mawww/kakoune/blob/master/doc/design.asciidoc#unified-interactive-use-and-scripting) -- to have a 1-to-1 relationship between scripted editing and interactive editing -- be against this feature?

**Comment by andreyorst** [comment]
@maximbaz if you still interested in this issue, what do you think about such solution?

https://discuss.kakoune.com/t/switching-keyboard-layouts-in-insert-mode/553

**Comment by maximbaz** [comment]
It's a really interesting and neat solution, works well, added it to my kakrc 🙂 Do make a plugin out of it.

At first I thought it is a bit weird that kakoune would not respect current OS keyboard layout in the insert mode, but on the other hand I rarely write in Russian in kakoune, and this approach nicely solves preserving my normal-mode mappings - and also I can even add `$kak_opt_langmap` to my modeline! 🙂 

Will keep using this and will let you know if long-term I will experience some issues with this. Thanks again 😉

**Comment by andreyorst** [comment]
> Do make a plugin out of it.

@maximbaz [I did](https://github.com/andreyorst/langmap.kak)

**Comment by vbauerster** [comment]
I used to use [vim-xkbswitch](https://github.com/lyokha/vim-xkbswitch) plugin, which relied on external cli utility to switch keyboard layout. The principle is just to memorize current layout in normal mode right before switching to insert mode and then restoring it upon leaving insert mode.

**Comment by andreyorst** [comment]
> I used to use [vim-xkbswitch](https://github.com/lyokha/vim-xkbswitch) plugin, which relies on external cli utility to switch keyboard layout. The principle is just to memorize current layout in normal mode right before switching to insert mode and then restoring it upon leaving insert mode.

I often need these settings to be buffer local, as I can work on documentation that has both Russian and English variants, and I often write them side by side to keep them in sync. The way plugin currently works it is possible, since langmap is set in the context of the buffer. I agree that using cli utility to change layout would be more general solution, but I think it will be hard to make it buffer local. You'll need to track a lot of stuff

**Comment by maximbaz** [comment]
> @maximbaz I did

Thanks! Although it's not compatible with the latest stable kakoune yet, will switch to your plugin once kakoune cuts off a new version 😉 

> I used to use vim-xkbswitch plugin, which relies on external cli utility to switch keyboard layout. The principle is just to memorize current layout in normal mode right before switching to insert mode and then restoring it upon leaving insert mode.

For me this would probably be even better, because I don't really need buffer-local layouts, instead I would appreciate if my keyboard layout in kakoune was synchronized with the OS layout. But in the absence of that plugin I can live with buffer-local one 😉

**Comment by andreyorst** [comment]
> Thanks! Although it's not compatible with the latest stable kakoune yet, will switch to your plugin once kakoune cuts off a new version 

You probably can just comment out module related lines and it should work.

**Comment by alexeymuranov** [comment]
In Vim a was so frustrated with the inability to change layout without breaking keymaps, and with keymaps based on "mnemonics" (useless unless you stick to QWERTY) instead of keyboard positions, that i made a [plugin](https://github.com/alexeymuranov/vim-CrazyKeys) for my personal use to remap all keys and to allow to restore key-function mapping after switching layout. (I am using it, but it is a half-baked hacky mess.)  About using `langmap` in Vim, it does not solve the problem because of [this bug](https://github.com/vim/vim/issues/297).

By the way, do i understand correctly that in Kakoune the <kbd>Enter</kbd> key is unused in the normal mode? Wouldn't it be better to use it to start the command mode (in addition to or instead of <kbd>:</kbd>)? ;)
